hypothesis-python/src/hypothesis/internal/conjecture/choicetree.py

```diff
@@ -0,0 +1,145 @@
+        if no such element exist".
+        """
+        assert not self.__finished
+        node = self.__node_trail[-1]
+        if node.size is None:
+            node.size = len(values)
+            node.n = len(values)
+
+        assert node.size > 0 or len(values) == 0
+
+        depth = len(self.__choices)
+
+        if depth < len(self.__prefix):
+            i = self.__prefix[depth]
+            if i >= len(values):
+                i = 0
+        else:
+            i = 0
+
+        count = 0
+        while node.size > 0:
+            count += 1
+            assert count <= len(values)
+            if not node.children[i].exhausted:
+                v = values[i]
+                if condition(v):
+                    self.__choices.append(i)
+                    self.__node_trail.append(node.children[i])
+                    return v
+                else:
+                    node.children[i] = DeadNode
+                    node.size -= 1
+            i = (i + 1) % len(values)
+        raise DeadBranch()
+
+    def finish(self):
+        """Record the decisions made in the underlying tree and return
+        a prefix that can be used for the next Chooser to be used."""
+        self.__finished = True
+        assert len(self.__node_trail) == len(self.__choices) + 1
+
+        next_value = list(self.__choices)
+        if next_value:
+            next_value[-1] += 1
+            for i in hrange(len(next_value) - 1, -1, -1):
+                if next_value[i] >= self.__node_trail[i].n:
+                    next_value[i] = 0
+                    if i > 0:
+                        next_value[i - 1] += 1
+                else:
+                    break
+
+        self.__node_trail[-1].size = 0
+        while len(self.__node_trail) > 1 and self.__node_trail[-1].exhausted:
+            self.__node_trail.pop()
+            assert len(self.__node_trail) == len(self.__choices)
+            i = self.__choices.pop()
+            target = self.__node_trail[-1]
+            target.children[i] = DeadNode
+            target.size -= 1
+
+        while len(next_value) > 0 and next_value[-1] == 0:
+            next_value.pop()
+
+        return tuple(next_value)
+
+
+class ChoiceTree(object):
+    """Records sequences of choices made during shrinking so that we
+    can track what parts of a pass has run. Used to create Chooser
+    objects that are the main interface that a pass uses to make
+    decisions about what to do.
+    """
+
+    def __init__(self):
+        self.root = TreeNode()
+
+    @property
+    def exhausted(self):
+        return self.root.exhausted
+
+    def step(self, prefix, f):
+        assert not self.exhausted
+        chooser = Chooser(self, prefix)
+        try:
+            f(chooser)
+        except DeadBranch:
+            pass
+        return chooser.finish()
+
+
+class TreeNode(object):
+    def __init__(self):
+        self.children = defaultdict(TreeNode)
+        self.size = None
```

Comment:
Minor: I found `size` a bit confusing, since it seems to really be tracking the remaining number of untried choices at a point in the tree.